### ISSUE TITLE
service.alpha.kubernetes.io/tolerate-unready-endpoints misuses the Endpoints API object

### ISSUE THREAD

**Issue by kow3ns** [kubernetes/kubernetes#49239]
**BUG REPORT**: The [addEndpointSubset method](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L509-L528) of the endpoints controller is currently adding EndpointSubsets with Address that are unready base on the `service.alpha.kubernetes.io/tolerate-unready-endpoints` annotation. 

At a minimum those addresses need to be added as EndpointSubsets with UnreadyAddresses, and the decision on whether to add them should be made, based on the annotation or its equivalent field, in the default DNS implementation's [handleEndpointUpdate](https://github.com/kubernetes/dns/blob/master/pkg/dns/dns.go#L320-L381) and [generateRecordsforHeadlessService](https://github.com/kubernetes/dns/blob/master/pkg/dns/dns.go#L480-L520) methods.

We might also consider simply always creating SRV records without respect to readiness. That is, the DNS implementation could disregard readiness for the endpoints associated with a headless services when creating SRV records and only worry about readiness when creating A records. However, if we consider this approach, we might want to give careful thought to what we might break.

Before proceeding with the implementation of either, I'd like to get feedback (and maybe participation) from SIG Network. Particularly as some clusters may not use the kube-dns add-on. I'd like to fix this in the 1.8 - 1.9 time frame if possible.

 /kind bug

**Comment by kow3ns** [comment]
@bgrant0607 @smarterclayton @kargakis @janetkuo

**Comment by caseydavenport** [comment]
@kubernetes/sig-network-bugs

**Comment by m1093782566** [comment]
I assume statefulset want kube-dns create A record for headless service even if pod is unready?@bprashanth @foxish

**Comment by thockin** [comment]
Not sure I understand the bug report.  It seems to be doing exactly what it is supposed to - the user asked to "tolerate" unready endpoints, so it is.  "tolerate" is maybe not the best word here, but still..

**Comment by kow3ns** [comment]
@thockin @bgrant0607's primary concern is that we're misusing the API. We're adding an unready address to the ready Addresses of the EndpointSubsets to create a SRV record so that stateful Pods can find their peers in a StatefulSet. The API already lets us register the unready endpoints for the headless service in the EndpointSubset's NotReadyAddresses. We should be detecting the annotation/field in kube-dns and adding NotReadyAddresses there. The users desire should be realized by the DNS implementation. More context [here](#49061).

**Comment by smarterclayton** [comment]
The current annotation explicitly says "tell everyone these endpoints are
ready" (which has a specific meaning, i.e. put me in rotation), so that
clients don't have to know the difference.  It forces legacy clients to
behave correctly.  Making a client respect the bool and interpret unready
endpoints would not transparently work for legacy clients.

I think the latter behavior is more correct, but the only difference
between the two approaches is whether the author can force legacy clients
to do what they expect, vs hope and wait for clients to be updated.

**Comment by kow3ns** [comment]
@smarterclayton I think the correctness of the behavior with respect to the use of Addresses vs UnreadyAddresses is @bgrant0607's primary concern.

@thockin @smarterclayton @bgrant0607 @erictune I would like to propose the following.

1. As a first step, we keep the current behavior of the endpoints controller and promote the annotation to a TolerateUnreadyEndpoints field.
    1. This will be a breaking API change with respect to the alpha annotation, but it will not modify the current behavior of the endpoint controller or require modification of any DNS addons.
    1. If we don't promote the field to at least beta, we can't reasonably expect that any DNS addons should modify their behavior to use it.
1. We (the community) modify all DNS addons to use the TolerateUnreadyEndpoints field to decide if they will add EndpointSubsets in the UnreadyAddresses to HeadlessServices.
    1. As the endpoint controller has not yet been modified this code will go into production dark.
    1. It would, by design, be compatible with current endpoint controller implementation and the modification proposed below.
1. We modify the endpoint controller to ignore the TolerateUnreadyEndpoints field, and pass the responsibility of implementing the TolerateUnreadyEndpoints behavior to the DNS addons.
    1. This will be a breaking, behavioral change, so we'd have to give DNS addon implementers time to adapt.

Alternatively we could
1. Leave the annotation as an alpha annotation.
1. Require that all DNS providers always add the UnreadyAddresses of EndpointSubsets for Headless Services.
    1. We're just adding a SRV here so this will be safe provided that clients of the HeadlessService don't require that only ready Pods should be added to DNS.
1. When all of the DNS addons have been updated, remove the alpha annotation and modify the endpoint controller.

I prefer the first approach because:
1. It was previously decided to promote the annotation, and it was indicated to the community that this is what we would deliver.
1. It gives us time to update the DNS addons, while allowing us to promote the functionality in such a way that users can depend on its consistent behavior.
1. If we never do steps 2 and 3, we have a working solution that can be advanced to GA.
1. If we do have a StatefulSet that requires readiness as a precondition for being added to DNS we support it.

The approach does have drawbacks:
1. We have to live with the TolerateUnreadyEndpoints field for the lifetime of v1.Service.
1. The second approach is simpler provided we are willing to live with the idea that StatefulSet Pods have to be able to determine the readiness of their peers without the help of a DNS addon that respects readiness checks (this is true for the majority of what is available today).

While the second approach benefits from simplicity it  also has downsides:
1. As previously mentioned, if we add DNS entries for UnreadyAddresses the Pods of StatefulSets must not depend on readiness to be communicated via DNS. 
1. We've left the annotation in alpha so long that it has been cargo culted into every StatefulSet example, into the stable helm repo, and into production (granted StatefulSet is still beta). I don't see how continuing to leave this annotation as alpha will help us.
1. Its an all or nothing, break both the existing alpha annotation and current behavior concurrently change.  This does not break our guarantees given that the annotation (and therefore the behavior) can be considered to be alpha, but it's highly probable that we'd be shipping v1.StatefulSet with dependencies on an alpha annotation we intend to deprecate.

**Comment by bgrant0607** [comment]
@kow3ns We shouldn't promote the annotation to a field and then change the semantics afterward. 

Why not a PublishNotReadyAddresses field (to match the name of the Endpoints field), which DNS could use to make the decision? We could leave the existing alpha annotation as is until the consumers we care about transition to the new approach. 

The NotReadyAddresses field was added 2 years ago, in #13778.

**Comment by kow3ns** [comment]
@bgrant0607 That approach SGTM, we would have to
1. Ship the annotation in 1.8
1. Move all the consumers (DNS addons, helm charts, examples etc) to use the field 1.9 and deprecate the annotation.
1. Remove the annotation in 1.10.

The only downside is that we will continue to carry the annotation for another 2 releases, and the cargo cult is likely to grow.

**Comment by bgrant0607** [comment]
We should also announce that the annotation is deprecated in 1.8.

**Comment by thockin** [comment]
I don't think this feature is just about DNS.  There are other clients of Endpoints, and this behavior catches them all.  I ACK that it's a little unintuitive, but frankly, "UnreadyEndpoints" is weird too.  I'd rather we never added that, and we just added a bool or annotation indicating whether readiness is considered important or not.

TL;DR - I think the current behavior is as correct as this is going to get.

**Comment by bgrant0607** [comment]
Many discovery APIs (e.g., Consul's, Eureka) include readiness/health information as a first-class property. It would be a shame to lose that.

**Comment by smarterclayton** [comment]
Two different consumers of a service can care about different sets of
endpoints (I don't remember if that had been brought back up on this
thread).  Forcing me to create multiple sets of services is a bit tedious.

**Comment by johnbelamaric** [comment]
@smarterclayton I have a use case where a container has a control port and a service port. The service port isn't "ready" until something attaches to the control port, loads some data, and executes a "make ready" API call.

Right now, this is handled by using this annotation for the control port service and not using it for the service port service. This is fine for me because my control port client just has to handle the possible non-readiness of the control port. But in theory this sort of use case would imply per-service readiness checks rather than per-container readiness checks.

The "juice may not be worth the squeeze" for that one, though.

**Comment by cblecker** [comment]
@kow3ns is this still an issue? I believe it is.
/remove-lifecycle rotten

**Comment by bgrant0607** [comment]
We need to take this into account when we redesign the Endpoints API

**Comment by thockin** [comment]
@freehan Can you digest this issue, make sure new Endpoints work addresses it (or explicitly does not) and then close?

**Comment by thockin** [comment]
@freehan - are we happy with new API wrt this?

**Comment by freehan** [comment]
I am leaning towards the 1st proposal in https://github.com/kubernetes/kubernetes/issues/49239#issuecomment-319501404

That means the new EndpointSlice controller will just ignore the `tolerate-unready-endpoint` annotation. Essentially, this decouples the DNS and endpoints. 

However, I am not sure if there is use case for `tolerate-unready-endpoint` on the endpoint side.

**Comment by johnbelamaric** [comment]
I don't see how that proposal works for non-DNS related consumers of these addresses. For example, won't unready addresses be programmed for cluster IP services when the field/annotation is used? This means kube-proxy or any other consumer will now have to re-implement the "tolerate unready" logic.

**Comment by freehan** [comment]
It looks like the new `PublishNotReadyAddresses` field is explicitly for DNS. https://github.com/kubernetes/kubernetes/blob/v1.16.0-alpha.3/pkg/apis/core/types.go#L3432

Please correct me if I am wrong, the proposal is NOT to support "tolerate unready" on the kube-proxy side. Unless there is real use cases.

**Comment by johnbelamaric** [comment]
It already is supported in kube-proxy. It's done by the endpoints controller. By the time it gets to kube-proxy, the decision is already made and so kube-proxy doesn't do anything special (neither does CoreDNS for that matter):

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L614

**Comment by johnbelamaric** [comment]
That said, it doesn't mean we *have* to support it the same way in endpoint slide - it's a different API after all. But we need consciously decide to not support it, if that's the case.

**Comment by johnbelamaric** [comment]
The stateful set use case is one, but I at least also used this feature at one point for pods that had a control port and a data port (https://github.com/kubernetes/kubernetes/issues/49239#issuecomment-351087689).

I wonder if the implicit "a pod is ready means the endpoint is ready" is just not the right model. Something more akin to the pod readiness gates but for service endpoints would be more flexible. So, "endpoint is ready if it meets these conditions" - where "pod readiness" is one possible condition, but other pod conditions could be looked at as well.

That's obviously a more complicated solution, but would allow folks to define "readiness" for their endpoints however they need for their particular use case (with the default being the current endpoint readiness == pod readiness).

Maybe a follow-on feature, if needed.

**Comment by freehan** [comment]
Yes. Agreed. 

I was thinking extending Pod ReadinessGate to allow user to specify if a pod readiness gate can block endpoint readiness or not. That would allow user to achieve something like "endpoint is ready" but "pod is not ready" when "all containers are ready" but "pod readiness gate is not ready but not blocking endpoint readiness". 

I feel like the readiness life cycle has to anchor on pods for backward compatibility because workflows only consider pods and none considers endpoint readiness.

This would be a separate feature any way.

**Comment by robscott** [comment]
It was a bit tricky to find when Endpoints controller stopped using this annotation, but EndpointSlice controller stopped in 1.17. Both controllers now use PublishNotReadyAddresses exclusively and have for some time. Closing this out.

/close
